kani-compiler/src/kani_middle/metadata.rs

```diff
@@ -40,6 +41,46 @@ pub fn gen_proof_metadata(tcx: TyCtxt, instance: Instance, base_name: &Path) ->
     }
 }
 
+/// Collects contract and contract harness metadata.
+///
+/// For each function with contracts (or that is a target of a contract harness),
+/// construct a ContractedFunction object for it.
+pub fn gen_contracts_metadata(tcx: TyCtxt) -> Vec<ContractedFunction> {
+    // We work with stable_mir::CrateItem instead of stable_mir::Instance to include generic items
+    let crate_items: CrateItems = stable_mir::all_local_items();
+
+    let mut fn_to_data: HashMap<DefId, ContractedFunction> = HashMap::new();
+
+    for item in crate_items {
```

Comment:
I think I'll leave it as-is because in this case:

```rust
if let Some(cf) = fn_to_data.get_mut(&target_def_id) {
    cf.harnesses.push(function);
}
```

nothing gets inserted into the hashmap, and I think `fold` makes more sense when every iteration folds something into the accumulator.